Security: Log injection vulnerability in NetworkPlugin

## Description

The NetworkPlugin logs user-controlled input without consistent sanitization, creating a log injection vulnerability. Attackers could inject terminal control sequences or manipulate logs.

## Vulnerability Details

**File:** `plugins/network/src/NetworkPlugin.ts`
**Lines:** 410, 483, 549, 556, 581, 716

### Affected Log Statements

```typescript
// Line 410 - User message content logged directly
console.log(`[Network] Message received:`, message);

// Line 483 - Event type from network
console.log(`[Network] Event:`, eventType, data);

// Line 549 - Entity data from network
console.log(`[Network] Received entity:`, entityData);

// Line 556 - Component data from network
console.log(`[Network] Component update:`, componentData);

// Line 581 - Client ID from network
console.log(`[Network] Client disconnected:`, clientId);

// Line 716 - Error messages from network
console.error(`[Network] Error:`, error.message);
```

## Attack Scenarios

### 1. Terminal Control Sequence Injection
Attacker sends message with ANSI escape codes:
```
\x1b[2J\x1b[H  // Clears screen
\x1b[31m       // Changes text color
```

### 2. Log Manipulation
Inject newlines to create fake log entries:
```
"Valid message\n[ERROR] SYSTEM COMPROMISED\n"
```

### 3. Log Parsing Attacks
If logs are parsed by monitoring tools, could inject malicious data.

## Recommended Fix

Add input sanitization function:

```typescript
/**
 * Sanitize user input before logging to prevent log injection
 */
function sanitizeForLog(input: unknown): string {
    if (typeof input !== 'string') {
        return JSON.stringify(input);
    }
    
    return input
        .replace(/\r/g, '\r')
        .replace(/\n/g, '\n')
        .replace(/\t/g, '\t')
        .replace(/\x1b/g, '\x1b')  // Escape ANSI codes
        .replace(/[\x00-\x1F\x7F]/g, (char) => 
            `\x${char.charCodeAt(0).toString(16).padStart(2, '0')}`
        );
}

// Usage
console.log(`[Network] Message received:`, sanitizeForLog(message));
```

### Alternative: Use Structured Logging

```typescript
// Better approach - use structured logging library
this.logger.info('message_received', {
    clientId: sanitize(clientId),
    messageType: sanitize(message.type),
    timestamp: Date.now()
});
```

## Impact

**Severity:** Medium-High
- Log injection attacks
- Terminal manipulation
- Log file corruption
- Monitoring system evasion
- False log entries

## Mitigation Priority

1. **Immediate:** Add sanitization to all network input logging
2. **Short-term:** Consider structured logging library
3. **Long-term:** Implement comprehensive input validation at network boundary

## Severity
**High** - Security vulnerability with moderate exploitability

## Related Issues
- Identified in nightly code review #245 (issue H4)
- Part of network plugin security audit

--- a/.changeset/fix-log-injection.md
+++ b/.changeset/fix-log-injection.md
@@ -1,0 +1,12 @@
+---
+"@orion-ecs/core": patch
+"@orion-ecs/plugin-api": patch
+"@orion-ecs/network": patch
+---
+
+Fix log injection vulnerability with centralized Logger
+
+- Add Logger interface to @orion-ecs/plugin-api for secure, structured logging
+- Create EngineLogger implementation with automatic sanitization of ANSI escape sequences and control characters
+- Update NetworkPlugin to use centralized logger via PluginContext
+- Plugins now import types from @orion-ecs/plugin-api for lighter dependencies

--- a/examples/multiplayer/bouncy-box/client.html
+++ b/examples/multiplayer/bouncy-box/client.html
@@ -224,11 +224,32 @@
     }
 
     /**
-     * Sanitize string for safe logging (remove newlines to prevent log injection)
+     * Sanitize string for safe logging to prevent log injection attacks.
+     * Removes newlines, ANSI escape sequences, and control characters.
      */
     function sanitizeLog(str) {
       if (typeof str !== 'string') return String(str);
-      return str.replace(/[\r\n]/g, '');
+      let result = '';
+      for (let i = 0; i < str.length; i++) {
+        const code = str.charCodeAt(i);
+        // Skip ANSI escape sequences (ESC [ ... letter)
+        if (code === 0x1b && str[i + 1] === '[') {
+          let j = i + 2;
+          while (j < str.length && ((str[j] >= '0' && str[j] <= '9') || str[j] === ';')) {
+            j++;
+          }
+          if (j < str.length && str[j] >= 'A' && str[j] <= 'z') {
+            i = j;
+            continue;
+          }
+        }
+        // Skip control characters (0x00-0x1f and 0x7f)
+        if (code < 0x20 || code === 0x7f) {
+          continue;
+        }
+        result += str[i];
+      }
+      return result;
     }
 
     /**

--- a/examples/multiplayer/bouncy-box/server.ts
+++ b/examples/multiplayer/bouncy-box/server.ts
@@ -37,11 +37,32 @@ import {
 // ============================================================================
 
 /**
- * Sanitize string for safe logging (remove newlines to prevent log injection)
+ * Sanitize string for safe logging to prevent log injection attacks.
+ * Removes newlines, ANSI escape sequences, and control characters.
  */
 function sanitizeLog(str: string): string {
     if (typeof str !== 'string') return String(str);
-    return str.replace(/[\r\n]/g, '');
+    let result = '';
+    for (let i = 0; i < str.length; i++) {
+        const code = str.charCodeAt(i);
+        // Skip ANSI escape sequences (ESC [ ... letter)
+        if (code === 0x1b && str[i + 1] === '[') {
+            let j = i + 2;
+            while (j < str.length && ((str[j] >= '0' && str[j] <= '9') || str[j] === ';')) {
+                j++;
+            }
+            if (j < str.length && str[j] >= 'A' && str[j] <= 'z') {
+                i = j;
+                continue;
+            }
+        }
+        // Skip control characters (0x00-0x1f and 0x7f)
+        if (code < 0x20 || code === 0x7f) {
+            continue;
+        }
+        result += str[i];
+    }
+    return result;
 }
 
 // ============================================================================

--- a/packages/core/src/definitions.ts
+++ b/packages/core/src/definitions.ts
@@ -9,7 +9,10 @@
  */
 
 // Import plugin types from @orion-ecs/plugin-api for local use
-import type { SystemMessage as PluginApiSystemMessage } from '@orion-ecs/plugin-api';
+import type {
+    Logger as PluginApiLogger,
+    SystemMessage as PluginApiSystemMessage,
+} from '@orion-ecs/plugin-api';
 
 // Re-export plugin types from @orion-ecs/plugin-api for backward compatibility
 // Plugin authors can import directly from @orion-ecs/plugin-api for a lighter dependency
@@ -17,11 +20,14 @@ export type {
     EnginePlugin,
     ExtractPluginExtensions,
     InstalledPlugin,
+    Logger,
+    LogLevel,
     SystemMessage,
 } from '@orion-ecs/plugin-api';
 
-// Alias for local use within this file
+// Aliases for local use within this file
 type SystemMessage = PluginApiSystemMessage;
+type Logger = PluginApiLogger;
 
 /**
  * Type representing a component class constructor.
@@ -770,6 +776,9 @@ export interface PluginContext {
 
     // Allow plugins to extend the engine with custom APIs
     extend<T extends object>(extensionName: string, api: T): void;
+
+    // Logger for secure, structured logging
+    logger: Logger;
 
     // Get engine instance for advanced use cases
     getEngine(): any; // Engine

--- a/packages/core/src/engine.ts
+++ b/packages/core/src/engine.ts
@@ -27,6 +27,7 @@ import type {
     EntityPrefab,
     ExtractPluginExtensions,
     InstalledPlugin,
+    Logger,
     MemoryStats,
     PluginContext,
     PoolStats,
@@ -35,6 +36,7 @@ import type {
     SystemProfile,
     SystemType,
 } from './definitions';
+import { EngineLogger } from './logger';
 import {
     ChangeTrackingManager,
     ComponentManager,
@@ -495,6 +497,9 @@ export class Engine {
     // Track if engine has been destroyed
     private _isDestroyed: boolean = false;
 
+    // Logger instance for the engine
+    private _logger: EngineLogger;
+
     constructor(
         private entityManager: EntityManager,
         private componentManager: ComponentManager,
@@ -509,6 +514,8 @@ export class Engine {
         private debugMode: boolean,
         private profilingEnabled: boolean = true
     ) {
+        // Initialize logger
+        this._logger = new EngineLogger({ debugEnabled: debugMode });
         // Subscribe to entity changes to update queries
         // Store unsubscribers for proper cleanup on destroy
         this.engineEventUnsubscribers.push(
@@ -1590,6 +1597,29 @@ export class Engine {
         };
     }
 
+    // ========== Logging ==========
+
+    /**
+     * Get the engine logger for secure, structured logging.
+     *
+     * The logger automatically sanitizes string arguments to prevent log injection
+     * attacks by removing ANSI escape sequences and control characters.
+     *
+     * @returns The engine logger instance
+     *
+     * @example
+     * ```typescript
+     * const logger = engine.logger.withTag('MySystem');
+     * logger.debug('Processing entity:', entity.name);
+     * logger.info('System initialized');
+     * logger.warn('Performance warning:', metrics);
+     * logger.error('Failed to process:', error);
+     * ```
+     */
+    get logger(): Logger {
+        return this._logger;
+    }
+
     // ========== Change Tracking ==========
 
     /**
@@ -2217,6 +2247,7 @@ export class Engine {
                 // Dynamically add extension to engine instance
                 (this as any)[extensionName] = api;
             },
+            logger: this._logger,
             getEngine: (): any => {
                 return this;
             },

--- a/packages/core/src/index.ts
+++ b/packages/core/src/index.ts
@@ -61,6 +61,8 @@ export type {
     EventTypes,
     ExtractPluginExtensions,
     InstalledPlugin,
+    Logger,
+    LogLevel,
     MemoryStats,
     ParentChangedEvent,
     ParentChangedListener,
@@ -78,6 +80,8 @@ export type {
 } from './definitions';
 // Export the new Engine and Builder
 export { Engine, EngineBuilder } from './engine';
+// Export logger utilities
+export { EngineLogger, sanitizeLogString } from './logger';
 // Export managers
 export {
     ChangeTrackingManager,

--- a/packages/core/src/logger.ts
+++ b/packages/core/src/logger.ts
@@ -1,0 +1,177 @@
+/**
+ * Logger Implementation for OrionECS
+ *
+ * Provides secure, structured logging with automatic sanitization
+ * to prevent log injection attacks.
+ *
+ * @packageDocumentation
+ * @module Logger
+ */
+
+import type { Logger, LogLevel } from './definitions';
+
+/**
+ * Sanitize a string to prevent log injection attacks.
+ *
+ * Removes:
+ * - ANSI escape sequences (terminal manipulation prevention)
+ * - Control characters (0x00-0x1F and 0x7F)
+ *
+ * @param str - The string to sanitize
+ * @returns The sanitized string
+ */
+export function sanitizeLogString(str: string): string {
+    if (typeof str !== 'string') return String(str);
+    let result = '';
+    for (let i = 0; i < str.length; i++) {
+        const code = str.charCodeAt(i);
+        // Skip ANSI escape sequences (ESC [ ... letter)
+        if (code === 0x1b && str[i + 1] === '[') {
+            let j = i + 2;
+            while (j < str.length && ((str[j] >= '0' && str[j] <= '9') || str[j] === ';')) {
+                j++;
+            }
+            if (j < str.length && str[j] >= 'A' && str[j] <= 'z') {
+                i = j;
+                continue;
+            }
+        }
+        // Skip control characters (0x00-0x1f and 0x7f)
+        if (code < 0x20 || code === 0x7f) {
+            continue;
+        }
+        result += str[i];
+    }
+    return result;
+}
+
+/**
+ * Sanitize log arguments, converting objects to strings as needed.
+ *
+ * @param args - The arguments to sanitize
+ * @returns Sanitized arguments
+ */
+function sanitizeArgs(args: unknown[]): unknown[] {
+    return args.map((arg) => {
+        if (typeof arg === 'string') {
+            return sanitizeLogString(arg);
+        }
+        if (arg instanceof Error) {
+            return sanitizeLogString(arg.message);
+        }
+        return arg;
+    });
+}
+
+/**
+ * Configuration options for the EngineLogger.
+ */
+export interface EngineLoggerOptions {
+    /** Enable debug-level logging */
+    debugEnabled?: boolean;
+    /** Minimum log level to output */
+    minLevel?: LogLevel;
+}
+
+/**
+ * Log level priority map for filtering.
+ */
+const LOG_LEVEL_PRIORITY: Record<LogLevel, number> = {
+    debug: 0,
+    info: 1,
+    warn: 2,
+    error: 3,
+};
+
+/**
+ * Engine logger implementation with automatic sanitization.
+ *
+ * This class implements the Logger interface and provides secure logging
+ * with ANSI escape sequence and control character filtering.
+ */
+export class EngineLogger implements Logger {
+    private tag: string;
+    private debugEnabled: boolean;
+    private minLevel: LogLevel;
+
+    constructor(options: EngineLoggerOptions = {}, tag: string = '') {
+        this.tag = tag;
+        this.debugEnabled = options.debugEnabled ?? false;
+        this.minLevel = options.minLevel ?? 'info';
+    }
+
+    /**
+     * Format the log prefix with tag if present.
+     */
+    private formatPrefix(): string {
+        return this.tag ? `[${this.tag}]` : '[ECS]';
+    }
+
+    /**
+     * Check if a log level should be output.
+     */
+    isEnabled(level: LogLevel): boolean {
+        if (level === 'debug' && !this.debugEnabled) {
+            return false;
+        }
+        return LOG_LEVEL_PRIORITY[level] >= LOG_LEVEL_PRIORITY[this.minLevel];
+    }
+
+    /**
+     * Log a debug message (only when debug mode is enabled).
+     */
+    debug(...args: unknown[]): void {
+        if (!this.isEnabled('debug')) return;
+        console.log(this.formatPrefix(), ...sanitizeArgs(args));
+    }
+
+    /**
+     * Log an informational message.
+     */
+    info(...args: unknown[]): void {
+        if (!this.isEnabled('info')) return;
+        console.log(this.formatPrefix(), ...sanitizeArgs(args));
+    }
+
+    /**
+     * Log a warning message.
+     */
+    warn(...args: unknown[]): void {
+        if (!this.isEnabled('warn')) return;
+        console.warn(this.formatPrefix(), ...sanitizeArgs(args));
+    }
+
+    /**
+     * Log an error message.
+     */
+    error(...args: unknown[]): void {
+        if (!this.isEnabled('error')) return;
+        console.error(this.formatPrefix(), ...sanitizeArgs(args));
+    }
+
+    /**
+     * Create a new logger instance with a tag prefix.
+     */
+    withTag(tag: string): Logger {
+        const newTag = this.tag ? `${this.tag}:${tag}` : tag;
+        const logger = new EngineLogger(
+            { debugEnabled: this.debugEnabled, minLevel: this.minLevel },
+            newTag
+        );
+        return logger;
+    }
+
+    /**
+     * Update the debug enabled state.
+     */
+    setDebugEnabled(enabled: boolean): void {
+        this.debugEnabled = enabled;
+    }
+
+    /**
+     * Update the minimum log level.
+     */
+    setMinLevel(level: LogLevel): void {
+        this.minLevel = level;
+    }
+}

--- a/packages/plugin-api/src/index.ts
+++ b/packages/plugin-api/src/index.ts
@@ -10,6 +10,111 @@
  */
 
 // =============================================================================
+// Logging Types
+// =============================================================================
+
+/**
+ * Log level for filtering log output.
+ *
+ * @public
+ */
+export type LogLevel = 'debug' | 'info' | 'warn' | 'error';
+
+/**
+ * Logger interface for structured logging with automatic sanitization.
+ *
+ * The logger provides a centralized, secure logging mechanism that:
+ * - Automatically sanitizes user-controlled input to prevent log injection attacks
+ * - Supports structured logging with tags/namespaces
+ * - Provides variadic methods for different log levels
+ *
+ * @remarks
+ * All string arguments passed to the logger are automatically sanitized to remove:
+ * - ANSI escape sequences (terminal manipulation prevention)
+ * - Control characters (0x00-0x1F and 0x7F)
+ *
+ * @example
+ * ```typescript
+ * class MyPlugin implements EnginePlugin {
+ *   install(context: PluginContext): void {
+ *     const logger = context.logger.withTag('MyPlugin');
+ *
+ *     logger.debug('Plugin initializing...');
+ *     logger.info('Connected to server:', serverName);
+ *     logger.warn('Connection slow, latency:', latencyMs, 'ms');
+ *     logger.error('Failed to connect:', error);
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
+export interface Logger {
+    /**
+     * Log a debug message (only visible when debug mode is enabled).
+     *
+     * @param args - Values to log (strings are automatically sanitized)
+     */
+    debug(...args: unknown[]): void;
+
+    /**
+     * Log an informational message.
+     *
+     * @param args - Values to log (strings are automatically sanitized)
+     */
+    info(...args: unknown[]): void;
+
+    /**
+     * Log a warning message.
+     *
+     * @param args - Values to log (strings are automatically sanitized)
+     */
+    warn(...args: unknown[]): void;
+
+    /**
+     * Log an error message.
+     *
+     * @param args - Values to log (strings are automatically sanitized)
+     */
+    error(...args: unknown[]): void;
+
+    /**
+     * Create a new logger instance with a tag prefix.
+     *
+     * Tags help identify the source of log messages, making it easier
+     * to filter and debug specific components.
+     *
+     * @param tag - Tag to prefix log messages with (e.g., plugin name)
+     * @returns A new Logger instance with the tag applied
+     *
+     * @example
+     * ```typescript
+     * const logger = context.logger.withTag('NetworkPlugin');
+     * logger.info('Connected'); // Output: [NetworkPlugin] Connected
+     * ```
+     */
+    withTag(tag: string): Logger;
+
+    /**
+     * Check if a log level is enabled.
+     *
+     * Useful for avoiding expensive string operations when the log
+     * level is disabled.
+     *
+     * @param level - The log level to check
+     * @returns True if the log level is enabled
+     *
+     * @example
+     * ```typescript
+     * if (logger.isEnabled('debug')) {
+     *   logger.debug('Expensive computation:', computeDebugInfo());
+     * }
+     * ```
+     */
+    isEnabled(level: LogLevel): boolean;
+}
+
+// =============================================================================
 // Entity Types
 // =============================================================================
 
@@ -508,6 +613,29 @@ export interface PluginContext {
      * ```
      */
     extend<T extends object>(extensionName: string, api: T): void;
+
+    // =========================================================================
+    // Logging
+    // =========================================================================
+
+    /**
+     * Logger instance for secure, structured logging.
+     *
+     * The logger automatically sanitizes string arguments to prevent log injection
+     * attacks. Use `logger.withTag()` to create a tagged logger for your plugin.
+     *
+     * @example
+     * ```typescript
+     * class MyPlugin implements EnginePlugin {
+     *   install(context: PluginContext): void {
+     *     const logger = context.logger.withTag('MyPlugin');
+     *     logger.info('Plugin initialized');
+     *     logger.debug('Debug details:', someValue);
+     *   }
+     * }
+     * ```
+     */
+    logger: Logger;
 
     // =========================================================================
     // Advanced

--- a/plugins/network/src/NetworkPlugin.ts
+++ b/plugins/network/src/NetworkPlugin.ts
@@ -40,7 +40,7 @@
  * ```
  */
 
-import type { EnginePlugin, EntityDef, PluginContext } from '@orion-ecs/core';
+import type { EnginePlugin, EntityDef, Logger, PluginContext } from '@orion-ecs/plugin-api';
 import {
     ClientInputState,
     InputBuffer,
@@ -175,6 +175,7 @@ export class NetworkPlugin implements EnginePlugin {
     readonly config: NetworkConfig;
 
     private context: PluginContext | null = null;
+    private logger: Logger | null = null;
     private networkAPI: NetworkAPI;
     private callbacks: NetworkEventCallbacks;
 
@@ -214,6 +215,7 @@ export class NetworkPlugin implements EnginePlugin {
 
     install(context: PluginContext): void {
         this.context = context;
+        this.logger = context.logger.withTag(`NetworkPlugin:${this.role}`);
         this.serverStartTime = Date.now();
         this.lastUpdateTime = this.serverStartTime;
 
@@ -276,7 +278,7 @@ export class NetworkPlugin implements EnginePlugin {
 
         this.transport.onConnect((connectionId) => {
             if (this.role === 'server') {
-                this.log(`Client connected: ${this.sanitizeLog(connectionId)}`);
+                this.log('Client connected:', connectionId);
             } else {
                 this.log('Connected to server');
                 if (this.callbacks.onConnected) {
@@ -289,7 +291,7 @@ export class NetworkPlugin implements EnginePlugin {
             if (this.role === 'server') {
                 this.handleClientDisconnect(connectionId, reason);
             } else {
-                this.log(`Disconnected: ${this.sanitizeLog(reason || '')}`);
+                this.log('Disconnected:', reason || '');
                 if (this.callbacks.onDisconnected) {
                     this.callbacks.onDisconnected(reason);
                 }
@@ -407,7 +409,7 @@ export class NetworkPlugin implements EnginePlugin {
             },
         });
 
-        this.log(`Player ${this.sanitizeLog(playerName)} joined (${this.clients.size} players)`);
+        this.log('Player', playerName, 'joined (', this.clients.size, 'players)');
 
         if (this.callbacks.onPlayerJoin) {
             this.callbacks.onPlayerJoin(clientId, playerName);
@@ -479,9 +481,7 @@ export class NetworkPlugin implements EnginePlugin {
             data: { clientId },
         });
 
-        this.log(
-            `Player ${this.sanitizeLog(connection.playerName)} left: ${this.sanitizeLog(reason || '')}`
-        );
+        this.log('Player', connection.playerName, 'left:', reason || '');
 
         if (this.callbacks.onPlayerLeave) {
             this.callbacks.onPlayerLeave(clientId);
@@ -546,14 +546,14 @@ export class NetworkPlugin implements EnginePlugin {
                 break;
 
             case 'player_joined':
-                this.log(`Player ${this.sanitizeLog(message.data.playerName)} joined`);
+                this.log('Player', message.data.playerName, 'joined');
                 if (this.callbacks.onPlayerJoin) {
                     this.callbacks.onPlayerJoin(message.data.clientId, message.data.playerName);
                 }
                 break;
 
             case 'player_left':
-                this.log(`Player ${this.sanitizeLog(message.data.clientId)} left`);
+                this.log('Player', message.data.clientId, 'left');
                 if (this.callbacks.onPlayerLeave) {
                     this.callbacks.onPlayerLeave(message.data.clientId);
                 }
@@ -577,9 +577,7 @@ export class NetworkPlugin implements EnginePlugin {
             ).setConnectionId(data.clientId);
         }
 
-        this.log(
-            `Joined as ${this.sanitizeLog(data.clientId)}, entity: ${this.sanitizeLog(data.networkEntityId)}`
-        );
+        this.log('Joined as', data.clientId, ', entity:', data.networkEntityId);
     }
 
     private handleWorldSnapshot(data: {
@@ -713,7 +711,7 @@ export class NetworkPlugin implements EnginePlugin {
 
         entity.queueFree();
         this.networkEntities.delete(networkEntityId);
-        this.log(`Entity destroyed: ${this.sanitizeLog(networkEntityId)}`);
+        this.log('Entity destroyed:', networkEntityId);
     }
 
     private createEntityFromSnapshot(data: SerializedNetworkEntity): EntityDef | undefined {
@@ -1086,17 +1084,13 @@ export class NetworkPlugin implements EnginePlugin {
     // Utilities
     // ============================================================================
 
-    private log(...args: unknown[]): void {
-        if (this.config.debug) {
-            console.log(`[NetworkPlugin:${this.role}]`, ...args);
-        }
-    }
-
     /**
-     * Sanitize string for safe logging (remove newlines to prevent log injection)
+     * Log a debug message using the engine's secure logger.
+     * The logger automatically sanitizes strings to prevent log injection attacks.
      */
-    private sanitizeLog(str: string): string {
-        if (typeof str !== 'string') return String(str);
-        return str.replace(/[\r\n]/g, '');
+    private log(...args: unknown[]): void {
+        if (this.config.debug && this.logger) {
+            this.logger.debug(...args);
+        }
     }
 }

--- a/plugins/network/src/transports/WebSocketClientTransport.ts
+++ b/plugins/network/src/transports/WebSocketClientTransport.ts
@@ -393,10 +393,32 @@ export class WebSocketClientTransport implements ClientTransport {
     }
 
     /**
-     * Sanitize string for safe logging (remove newlines to prevent log injection)
+     * Sanitize string for safe logging to prevent log injection attacks.
+     * Removes newlines, ANSI escape sequences, and control characters.
      */
     private sanitizeLog(str: string): string {
-        return str.replace(/[\r\n]/g, '');
+        if (typeof str !== 'string') return String(str);
+        let result = '';
+        for (let i = 0; i < str.length; i++) {
+            const code = str.charCodeAt(i);
+            // Skip ANSI escape sequences (ESC [ ... letter)
+            if (code === 0x1b && str[i + 1] === '[') {
+                let j = i + 2;
+                while (j < str.length && ((str[j] >= '0' && str[j] <= '9') || str[j] === ';')) {
+                    j++;
+                }
+                if (j < str.length && str[j] >= 'A' && str[j] <= 'z') {
+                    i = j;
+                    continue;
+                }
+            }
+            // Skip control characters (0x00-0x1f and 0x7f)
+            if (code < 0x20 || code === 0x7f) {
+                continue;
+            }
+            result += str[i];
+        }
+        return result;
     }
 }
 

--- a/plugins/network/src/transports/WebSocketServerTransport.ts
+++ b/plugins/network/src/transports/WebSocketServerTransport.ts
@@ -293,11 +293,32 @@ export class WebSocketServerTransport implements ServerTransport {
     }
 
     /**
-     * Sanitize string for safe logging (remove newlines to prevent log injection)
+     * Sanitize string for safe logging to prevent log injection attacks.
+     * Removes newlines, ANSI escape sequences, and control characters.
      */
     private sanitizeLog(str: string): string {
         if (typeof str !== 'string') return String(str);
-        return str.replace(/[\r\n]/g, '');
+        let result = '';
+        for (let i = 0; i < str.length; i++) {
+            const code = str.charCodeAt(i);
+            // Skip ANSI escape sequences (ESC [ ... letter)
+            if (code === 0x1b && str[i + 1] === '[') {
+                let j = i + 2;
+                while (j < str.length && ((str[j] >= '0' && str[j] <= '9') || str[j] === ';')) {
+                    j++;
+                }
+                if (j < str.length && str[j] >= 'A' && str[j] <= 'z') {
+                    i = j;
+                    continue;
+                }
+            }
+            // Skip control characters (0x00-0x1f and 0x7f)
+            if (code < 0x20 || code === 0x7f) {
+                continue;
+            }
+            result += str[i];
+        }
+        return result;
     }
 }
 

